File: tests/omnia/data/cache/SimpleDoubleCacherTest.java
package omnia.data.cache;

import static com.google.common.truth.Truth.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.function.DoubleSupplier;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)
public class SimpleDoubleCacherTest {

  @Test
  public void value_didReturnSuppliedValue() {
    double testValue = 132.0;
    CachedDouble testSubject = new SimpleDoubleCacher(() -> testValue);

    assertThat(testSubject.value()).isEqualTo(testValue);
  }

  @Test
  public void value_twice_didReturnSuppliedValue() {
    double testValue = 132.0;
    CachedDouble testSubject = new SimpleDoubleCacher(() -> testValue);

    testSubject.value();

    assertThat(testSubject.value()).isEqualTo(testValue);
  }

  @Test
  public void new_didNotInvokeSupplier() {
    DoubleSupplier supplier = setUpMockSupplier();
    new SimpleDoubleCacher(supplier);

    verify(supplier, never()).getAsDouble();
  }

  @Test
  public void invalidate_didNotSupplier() {
    DoubleSupplier supplier = setUpMockSupplier();
    CachedDouble testSubject = new SimpleDoubleCacher(supplier);

    testSubject.invalidate();

    verify(supplier, never()).getAsDouble();
  }

  @Test
  public void value_didInvokeSupplier() {
    DoubleSupplier supplier = setUpMockSupplier();
    CachedDouble testSubject = new SimpleDoubleCacher(supplier);

    testSubject.value();

    verify(supplier).getAsDouble();
  }

  @Test
  public void value_didInvokeSupplierOnce() {
    DoubleSupplier supplier = setUpMockSupplier();
    CachedDouble testSubject = new SimpleDoubleCacher(supplier);

    testSubject.value();
    testSubject.value();

    verify(supplier, times(1)).getAsDouble();
  }

  @Test
  public void value_thenInvalidate_thenValue_didInvokeSupplierTwice() {
    DoubleSupplier supplier = setUpMockSupplier();
    CachedDouble testSubject = new SimpleDoubleCacher(supplier);

    testSubject.value();
    testSubject.invalidate();
    testSubject.value();

    verify(supplier, times(2)).getAsDouble();
  }

  @Test
  public void new_withNullSupplier_didThrowException() {
    assertThrows(NullPointerException.class, () -> new SimpleDoubleCacher(null));
  }

  private static DoubleSupplier setUpMockSupplier() {
    DoubleSupplier supplier = mock(DoubleSupplier.class);
    when(supplier.getAsDouble()).thenReturn(132.0);
    return supplier;
  }
}

File: tests/omnia/data/cache/SimpleDoubleCacherTest.kt
package omnia.data.cache

import com.google.common.truth.Truth.assertThat
import kotlin.test.Test

import org.mockito.Mockito
import org.mockito.kotlin.mock

class SimpleDoubleCacherTest {

  @Test
  fun value_didReturnSuppliedValue() {
    val testValue = 132.0
    val testSubject: CachedDouble = SimpleDoubleCacher { testValue }
    assertThat(testSubject.value()).isEqualTo(testValue)
  }

  @Test
  fun value_twice_didReturnSuppliedValue() {
    val testValue = 132.0
    val testSubject: CachedDouble = SimpleDoubleCacher { testValue }
    testSubject.value()
    assertThat(testSubject.value()).isEqualTo(testValue)
  }

  @Test
  fun new_didNotInvokeSupplier() {
    val supplier = setUpMockSupplier()
    SimpleDoubleCacher(supplier)
    Mockito.verify(supplier, Mockito.never()).invoke()
  }

  @Test
  fun invalidate_didNotSupplier() {
    val supplier = setUpMockSupplier()
    val testSubject: CachedDouble = SimpleDoubleCacher(supplier)
    testSubject.invalidate()
    Mockito.verify(supplier, Mockito.never()).invoke()
  }

  @Test
  fun value_didInvokeSupplier() {
    val supplier = setUpMockSupplier()
    val testSubject: CachedDouble = SimpleDoubleCacher(supplier)
    testSubject.value()
    Mockito.verify(supplier).invoke()
  }

  @Test
  fun value_didInvokeSupplierOnce() {
    val supplier = setUpMockSupplier()
    val testSubject: CachedDouble = SimpleDoubleCacher(supplier)
    testSubject.value()
    testSubject.value()
    Mockito.verify(supplier, Mockito.times(1)).invoke()
  }

  @Test
  fun value_thenInvalidate_thenValue_didInvokeSupplierTwice() {
    val supplier = setUpMockSupplier()
    val testSubject: CachedDouble = SimpleDoubleCacher(supplier)
    testSubject.value()
    testSubject.invalidate()
    testSubject.value()
    Mockito.verify(supplier, Mockito.times(2)).invoke()
  }

  companion object {

    private fun setUpMockSupplier(): () -> Double {
      val supplier = mock<() -> Double>()
      Mockito.`when`(supplier()).thenReturn(132.0)
      return supplier
    }
  }
}